libs/tour/src/lib/services/tour-service/tour.service.ts

```diff
@@ -0,0 +1,486 @@
+import { ComponentRef, Inject, Injectable, PLATFORM_ID, Type } from '@angular/core';
+import {
+	BehaviorSubject,
+	Observable,
+	Subject,
+	auditTime,
+	concatMap,
+	distinctUntilChanged,
+	filter,
+	of,
+	switchMap,
+	take,
+	takeUntil,
+	tap,
+} from 'rxjs';
+import { ConnectedPosition, Overlay, OverlayConfig, OverlayRef } from '@angular/cdk/overlay';
+import { ComponentPortal } from '@angular/cdk/portal';
+import { isPlatformBrowser } from '@angular/common';
+
+import { NgxTourStepComponent } from '../../abstracts';
+import { NgxTourItemDirective } from '../../directives';
+import { NgxTourStepToken } from '../../tokens';
+import {
+	NgxTourDirection,
+	NgxTourAction,
+	NgxTourStep,
+	NgxTourInteraction,
+	NgxTourStepPosition,
+} from '../../types';
+import { elementIsVisibleInViewport } from '../../utils';
+
+/**
+ * A singleton service used to run help tours through an application.
+ */
+@Injectable({
+	providedIn: 'root',
+})
+export class NgxTourService {
+	/**
+	 * A record of registered step elements we wish to highlight
+	 */
+	private elements: Record<string, BehaviorSubject<NgxTourItemDirective>> = {};
+	/**
+	 * A record of positions to place the
+	 */
+	private readonly positionMap: Record<NgxTourStepPosition, ConnectedPosition> = {
+		below: { originX: 'start', originY: 'bottom', overlayX: 'start', overlayY: 'top' },
+		above: { originX: 'start', originY: 'top', overlayX: 'start', overlayY: 'bottom' },
+		left: { originX: 'start', originY: 'bottom', overlayX: 'end', overlayY: 'bottom' },
+		right: { originX: 'end', originY: 'center', overlayX: 'start', overlayY: 'bottom' },
+	};
+
+	/**
+	 * The currently active overlay
+	 */
+	private overlayRef: OverlayRef;
+
+	/**
+	 * The amount of steps of the current tour
+	 */
+	private amountOfSteps: number = 0;
+
+	/**
+	 * The current direction we're moving the tour in
+	 */
+	private currentDirection: NgxTourDirection;
+
+	/**
+	 * The currently active step in the tour as a subject
+	 */
+	private readonly currentStepSubject: BehaviorSubject<NgxTourStep> =
+		new BehaviorSubject<NgxTourStep>(undefined);
+
+	/**
+	 * The previously active step in the tour as a subject
+	 */
+	private readonly previousStepSubject: BehaviorSubject<NgxTourStep> =
+		new BehaviorSubject<NgxTourStep>(undefined);
+
+	/**
+	 * The index of the current step of the tour as a subject
+	 */
+	private readonly currentIndexSubject: BehaviorSubject<number> = new BehaviorSubject<number>(0);
+
+	/**
+	 * Whether the tour has started as a subject
+	 */
+	private readonly tourStartedSubject: Subject<void> = new Subject();
+
+	/**
+	 * Whether the tour has ended as a subject
+	 */
+	private readonly tourEndedSubject: Subject<void> = new Subject();
+
+	/**
+	 * The currently active tour as a subject
+	 */
+	private readonly currentTourSubject: BehaviorSubject<NgxTourStep[]> = new BehaviorSubject<
+		NgxTourStep[]
+	>(undefined);
```

Comment:
Might be better not to use an implicit `undefined` but use `null` instead?
(also applies to other code in the PR)

Reply:
I left these subjects as is, as these can later on be changed by other values; but in the other places I updated it to be null 👍